src/EntryPoint.sol

```diff
@@ -168,113 +142,125 @@ contract EntryPoint is EIP712, UUPSUpgradeable, Ownable, ReentrancyGuard {
+                revert(0x1c, 0x04)
+            }
+
+            let m := mload(0x40) // Grab the free memory pointer.
+            // Copy the encoded user op to the memory to be ready to pass to the self call.
+            calldatacopy(add(m, 0x20), encodedUserOp.offset, encodedUserOp.length)
+            let s := add(m, 0x1c) // Start of the calldata in memory to pass to the self call.
+            let n := add(encodedUserOp.length, 0x24) // Length of the calldata to the self call.
+
+            // To prevent griefing, we need to do two non-reverting gas-limited calls.
+            // Even if the verify and call fails, which the gas will be burned,
+            // the payment has already been made and can't be reverted.
+
+            // 1. Pay.
+            mstore(m, 0x1a3de5c3) // `_pay()`.
+            mstore(0x00, 0) // Zeroize the return slot.
+            let gCapped := xor(g, mul(xor(g, _PAYMENT_GAS_CAP), lt(_PAYMENT_GAS_CAP, g))) // `min`.
+            // Perform the gas-limited self call.
+            switch call(gCapped, address(), 0, s, n, 0x00, 0x20)
+            case 0 {
+                err := mload(0x00)
+                if iszero(returndatasize()) { err := shl(224, 0xbff2584f) } // `PaymentError()`.
+            }
+            default {
+                // Since the payment is a success, load the returned `paymentAmount`.
+                paymentAmount := mload(0x00)
+                let gUsed := sub(gStart, gas())
+                let gLeft := mul(sub(g, gUsed), gt(g, gUsed))
+                // 2. Verify and call.
+                mstore(m, 0xe235a92a) // `_verifyAndCall()`.
+                mstore(0x00, 0) // Zeroize the return slot.
+                // Perform the gas-limited self call.
+                if iszero(call(gLeft, address(), 0, s, n, 0x00, 0x20)) {
+                    err := mload(0x00)
+                    if iszero(returndatasize()) { err := shl(224, 0xad4db224) } // `VerifiedCallError()`.
+                }
+            }
+        }
+
+        // Refund strategy:
+        // `totalAmountOfGasToPayFor = gasUsedThusFar + _REFUND_GAS`.
+        // `paymentAmountForGas = paymentPerGas * totalAmountOfGasToPayFor`.
+        // If we have overpaid, then refund `paymentAmount - paymentAmountForGas`.
+
+        uint256 gUsed = Math.rawSub(gStart, gasleft());
+        uint256 paymentPerGas = u.paymentPerGas;
+        if (paymentPerGas == uint256(0)) paymentPerGas = type(uint256).max;
+        uint256 finalPaymentAmount = Math.min(
+            paymentAmount, Math.saturatingMul(paymentPerGas, Math.saturatingAdd(gUsed, _REFUND_GAS))
+        );
+        address paymentRecipient = u.paymentRecipient;
+        if (paymentRecipient == address(0)) paymentRecipient = address(this);
+        if (LibBit.and(finalPaymentAmount != 0, paymentRecipient != address(this))) {
+            TokenTransferLib.safeTransfer(u.paymentToken, paymentRecipient, finalPaymentAmount);
+        }
```

Comment:
I don't understand, given we are doing this stuff, why do we even call `_pay(u)` above? Let's just call `payEntrypoint(...)` here with `finalPaymentAmount` to `paymentRecipient` instead of first pulling the funds in from the account to the entrypoint, and then from the entrypoint to the payment recipient?